src/cio_websocket.c

```diff
@@ -0,0 +1,466 @@
+	if (err == CIO_SUCCESS) {
+		err = ws->close_timer.expires_from_now(&ws->close_timer, close_timeout_ns, close_timeout_handler, ws);
+		if (err == CIO_SUCCESS) {
+			send_frame(ws, (uint8_t *)&ws->close_status, sizeof(ws->close_status), CIO_WEBSOCKET_CLOSE_FRAME, close_frame_written);
+			return;
+		}
+	}
+
+	send_frame(ws, (uint8_t *)&ws->close_status, sizeof(ws->close_status), CIO_WEBSOCKET_CLOSE_FRAME, do_nothing);
+	close(ws);
+}
+
+static void handle_frame(struct cio_websocket *ws, uint8_t *data, uint64_t length)
+{
+	if (unlikely(ws->is_server && (ws->ws_flags.mask == 0))) {
+		// TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		return;
+	}
+
+	if (ws->ws_flags.mask != 0) {
+		cio_websocket_mask(data, length, ws->mask);
+	}
+
+	if (unlikely((ws->ws_flags.fin == 0) && (ws->ws_flags.opcode >= CIO_WEBSOCKET_CLOSE_FRAME))) {
+		// TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		return;
+	}
+
+	if (ws->ws_flags.fin == 0) {
+		if (ws->ws_flags.opcode != CIO_WEBSOCKET_CONTINUATION_FRAME) {
+			if (unlikely(ws->ws_flags.is_fragmented)) {
+				// TODO: log_err("Overwriting Opcode of unfinished fragmentation!");
+				// TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+				return;
+			}
+
+			ws->ws_flags.is_fragmented = 1;
+			ws->ws_flags.frag_opcode = ws->ws_flags.opcode;
+			ws->ws_flags.opcode = CIO_WEBSOCKET_CONTINUATION_FRAME;
+		} else {
+			if (unlikely(ws->ws_flags.rsv != 0)) {
+				// TODO:log_err("RSV-Bits must be 0 during continuation!");
+				// TODO:handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+				return;
+			}
+
+			if (unlikely(!(ws->ws_flags.is_fragmented))) {
+				//TODO: log_err("No start frame was send!");
+				//TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+				return;
+			}
+		}
+	}
+
+	if (unlikely(ws->ws_flags.is_fragmented && ((ws->ws_flags.opcode < CIO_WEBSOCKET_CLOSE_FRAME) && (ws->ws_flags.opcode > CIO_WEBSOCKET_CONTINUATION_FRAME)))) {
+		// log_err("Opcode during fragmentation must be 0x0!");
+		// handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		return;
+	}
+
+	bool last_frame = false;
+	unsigned int opcode;
+	if (ws->ws_flags.opcode == CIO_WEBSOCKET_CONTINUATION_FRAME) {
+		opcode = ws->ws_flags.frag_opcode;
+		if (unlikely(ws->ws_flags.fin ==1)) {
+			last_frame = true;
+		}
+	} else {
+		opcode = ws->ws_flags.opcode;
+	}
+
+	switch (opcode) {
+	case CIO_WEBSOCKET_BINARY_FRAME:
+		if (likely(ws->onbinaryframe_received != NULL)) {
+			ws->onbinaryframe_received(ws, data, length, last_frame);
+		} else {
+			// TODO: handle_error(s, WS_CLOSE_UNSUPPORTED);
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_TEXT_FRAME:
+		if (likely(ws->on_textframe_received != NULL)) {
+			ws->on_textframe_received(ws, (char *)data, length, last_frame);
+		} else {
+			// TODO: handle_error(s, WS_CLOSE_UNSUPPORTED);
+		}
+
+		break;
+
+	case CIO_WEBSOCKET_PING_FRAME:
+		if (unlikely(length > WS_SMALL_FRAME_SIZE)) {
+			//TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		} else {
+			// TODO: send_pong_frame(ws, data, length);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this block of commented out code. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3ACommentedCode)
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS1135)